--- a/target/linux/sunxi/config-5.10	2023-12-03 21:02:20.000000000 +0200
+++ b/target/linux/sunxi/config-5.10	2024-01-20 13:26:16.929438059 +0200
@@ -90,7 +90,8 @@
 CONFIG_CPU_FREQ_GOV_PERFORMANCE=y
 CONFIG_CPU_FREQ_GOV_POWERSAVE=y
 CONFIG_CPU_FREQ_GOV_USERSPACE=y
-CONFIG_CPU_FREQ_STAT=y
+# CONFIG_CPU_FREQ_STAT is not set
+CONFIG_CPU_FREQ_THERMAL=y
 CONFIG_CPU_HAS_ASID=y
 CONFIG_CPU_PABRT_V7=y
 CONFIG_CPU_PM=y
@@ -171,6 +172,7 @@
 CONFIG_FS_POSIX_ACL=y
 CONFIG_FW_CACHE=y
 CONFIG_FW_LOADER_PAGED_BUF=y
+CONFIG_GENERIC_ADC_THERMAL=y
 CONFIG_GENERIC_ALLOCATOR=y
 CONFIG_GENERIC_ARCH_TOPOLOGY=y
 CONFIG_GENERIC_BUG=y
@@ -225,6 +227,7 @@
 CONFIG_INITRAMFS_SOURCE=""
 CONFIG_INPUT=y
 CONFIG_INPUT_AXP20X_PEK=y
+CONFIG_INPUT_EVDEV=y
 CONFIG_INPUT_KEYBOARD=y
 CONFIG_INPUT_MOUSEDEV=y
 CONFIG_INPUT_MOUSEDEV_PSAUX=y
@@ -236,6 +239,25 @@
 CONFIG_IRQ_DOMAIN_HIERARCHY=y
 CONFIG_IRQ_FORCED_THREADING=y
 CONFIG_IRQ_WORK=y
+CONFIG_IR_GPIO_CIR=m
+CONFIG_IR_GPIO_TX=m
+CONFIG_IR_IMON_DECODER=m
+CONFIG_IR_JVC_DECODER=m
+CONFIG_IR_MCE_KBD_DECODER=m
+CONFIG_IR_NEC_DECODER=m
+# CONFIG_IR_PWM_TX is not set
+CONFIG_IR_RC5_DECODER=m
+CONFIG_IR_RC6_DECODER=m
+CONFIG_IR_RCMM_DECODER=m
+CONFIG_IR_SANYO_DECODER=m
+# CONFIG_IR_SERIAL is not set
+CONFIG_IR_SHARP_DECODER=m
+# CONFIG_IR_SIR is not set
+CONFIG_IR_SONY_DECODER=m
+# CONFIG_IR_SPI is not set
+CONFIG_IR_SUNXI=m
+# CONFIG_IR_TOY is not set
+CONFIG_IR_XMP_DECODER=m
 CONFIG_JBD2=y
 CONFIG_KALLSYMS=y
 CONFIG_KEYBOARD_SUN4I_LRADC=y
@@ -244,6 +266,7 @@
 CONFIG_LCD_PLATFORM=y
 CONFIG_LEDS_GPIO=y
 CONFIG_LIBFDT=y
+CONFIG_LIRC=y
 CONFIG_LOCK_DEBUGGING_SUPPORT=y
 CONFIG_LOCK_SPIN_ON_OWNER=y
 CONFIG_LOGO=y
@@ -368,6 +391,11 @@
 CONFIG_PWM_SUN4I=y
 CONFIG_PWM_SYSFS=y
 CONFIG_RATIONAL=y
+CONFIG_RC_CORE=m
+CONFIG_RC_DECODERS=y
+CONFIG_RC_DEVICES=y
+CONFIG_RC_MAP=m
+# CONFIG_RC_XBOX_DVD is not set
 CONFIG_REALTEK_PHY=y
 CONFIG_REGMAP=y
 CONFIG_REGMAP_I2C=y
@@ -388,6 +416,7 @@
 CONFIG_RWSEM_SPIN_ON_OWNER=y
 CONFIG_SATA_HOST=y
 CONFIG_SATA_PMP=y
+CONFIG_SCHED_THERMAL_PRESSURE=y
 CONFIG_SCSI=y
 CONFIG_SDIO_UART=y
 CONFIG_SECURITYFS=y
@@ -406,17 +435,19 @@
 CONFIG_SMP_ON_UP=y
 CONFIG_SND=y
 CONFIG_SND_COMPRESS_OFFLOAD=y
+CONFIG_SND_DMAENGINE_PCM=y
 CONFIG_SND_JACK=y
 CONFIG_SND_JACK_INPUT_DEV=y
 CONFIG_SND_PCM=y
-CONFIG_SND_SIMPLE_CARD=y
-CONFIG_SND_SIMPLE_CARD_UTILS=y
 CONFIG_SND_SOC=y
+CONFIG_SND_SOC_GENERIC_DMAENGINE_PCM=y
 CONFIG_SND_SOC_I2C_AND_SPI=y
-# CONFIG_SND_SUN4I_I2S is not set
-# CONFIG_SND_SUN4I_SPDIF is not set
-# CONFIG_SND_SUN8I_CODEC is not set
-# CONFIG_SND_SUN8I_CODEC_ANALOG is not set
+CONFIG_SND_SUN4I_CODEC=y
+CONFIG_SND_SUN4I_I2S=y
+CONFIG_SND_SUN4I_SPDIF=y
+CONFIG_SND_SUN8I_ADDA_PR_REGMAP=y
+CONFIG_SND_SUN8I_CODEC=y
+CONFIG_SND_SUN8I_CODEC_ANALOG=y
 CONFIG_SOUND=y
 CONFIG_SOUND_OSS_CORE=y
 CONFIG_SPARSE_IRQ=y
@@ -431,6 +462,7 @@
 # CONFIG_STMMAC_SELFTESTS is not set
 CONFIG_SUN4I_A10_CCU=y
 # CONFIG_SUN4I_EMAC is not set
+CONFIG_SUN4I_GPADC=y
 CONFIG_SUN4I_TIMER=y
 CONFIG_SUN5I_CCU=y
 CONFIG_SUN5I_HSTIMER=y
@@ -460,9 +492,12 @@
 CONFIG_THERMAL=y
 CONFIG_THERMAL_DEFAULT_GOV_STEP_WISE=y
 CONFIG_THERMAL_EMERGENCY_POWEROFF_DELAY_MS=0
+CONFIG_THERMAL_GOV_BANG_BANG=y
+CONFIG_THERMAL_GOV_FAIR_SHARE=y
 CONFIG_THERMAL_GOV_STEP_WISE=y
 CONFIG_THERMAL_HWMON=y
 CONFIG_THERMAL_OF=y
+CONFIG_THERMAL_WRITABLE_TRIPS=y
 CONFIG_TICK_CPU_ACCOUNTING=y
 CONFIG_TIMER_OF=y
 CONFIG_TIMER_PROBE=y
@@ -474,6 +509,7 @@
 CONFIG_UNCOMPRESS_INCLUDE="debug/uncompress.h"
 CONFIG_UNWINDER_ARM=y
 CONFIG_USB=y
+CONFIG_USB_ACM=y
 CONFIG_USB_ANNOUNCE_NEW_DEVICES=y
 # CONFIG_USB_AUDIO is not set
 CONFIG_USB_COMMON=y
@@ -481,13 +517,21 @@
 CONFIG_USB_DWC2_HOST=y
 CONFIG_USB_EHCI_HCD=y
 CONFIG_USB_EHCI_HCD_PLATFORM=y
+CONFIG_USB_ETH=y
+# CONFIG_USB_ETH_EEM is not set
+CONFIG_USB_ETH_RNDIS=y
+CONFIG_USB_F_ECM=y
+CONFIG_USB_F_RNDIS=y
+CONFIG_USB_F_SUBSET=y
 CONFIG_USB_GADGET=y
+CONFIG_USB_LIBCOMPOSITE=y
 CONFIG_USB_NET_DRIVERS=y
 CONFIG_USB_OHCI_HCD=y
 CONFIG_USB_OHCI_HCD_PLATFORM=y
 CONFIG_USB_ROLE_SWITCH=y
 CONFIG_USB_STORAGE=y
 CONFIG_USB_SUPPORT=y
+CONFIG_USB_U_ETHER=y
 CONFIG_USERIO=y
 CONFIG_USE_OF=y
 CONFIG_VFAT_FS=y
@@ -496,8 +540,6 @@
 CONFIG_VHOST=y
 CONFIG_VHOST_IOTLB=y
 CONFIG_VHOST_NET=y
-# CONFIG_VIDEO_SUN4I_CSI is not set
-# CONFIG_VIDEO_SUN6I_CSI is not set
 CONFIG_VM_EVENT_COUNTERS=y
 CONFIG_VT=y
 CONFIG_VT_CONSOLE=y
--- a/target/linux/sunxi/modules.mk	2023-12-03 21:02:20.000000000 +0200
+++ b/target/linux/sunxi/modules.mk	2024-01-20 13:26:16.929438059 +0200
@@ -23,15 +23,60 @@
 define KernelPackage/sunxi-ir
     SUBMENU:=$(OTHER_MENU)
     TITLE:=Sunxi SoC built-in IR support (A20)
-    DEPENDS:=@TARGET_sunxi +kmod-input-core
+    DEPENDS:=@TARGET_sunxi +kmod-input-core +v4l-utils +triggerhappy
     $(call AddDepends/rtc)
     KCONFIG:= \
 	CONFIG_MEDIA_SUPPORT=y \
 	CONFIG_MEDIA_RC_SUPPORT=y \
+	CONFIG_IR_SUNXI=m \
+	CONFIG_RC_CORE=m \
+	CONFIG_RC_DECODERS=y \
 	CONFIG_RC_DEVICES=y \
-	CONFIG_IR_SUNXI
-    FILES:=$(LINUX_DIR)/drivers/media/rc/sunxi-cir.ko
-    AUTOLOAD:=$(call AutoLoad,50,sunxi-cir)
+	CONFIG_RC_MAP=m \
+	CONFIG_IR_LIRC_CODEC=m \
+	CONFIG_BPF_LIRC_MODE2=y \
+	CONFIG_IR_NEC_DECODER=m \
+	CONFIG_IR_RC5_DECODER=m \
+	CONFIG_IR_RC6_DECODER=m \
+	CONFIG_IR_JVC_DECODER=m \
+	CONFIG_IR_SONY_DECODER=m \
+	CONFIG_IR_SANYO_DECODER=m \
+	CONFIG_IR_SHARP_DECODER=m \
+	CONFIG_IR_MCE_KBD_DECODER=m \
+	CONFIG_IR_XMP_DECODER=m \
+	CONFIG_IR_IMON_DECODER=m \
+	CONFIG_IR_RCMM_DECODER=n \
+	CONFIG_IR_IMON_RAW=m \
+	CONFIG_IR_SPI=m \
+	CONFIG_IR_PWM_TX=m \
+	CONFIG_IR_SERIAL=m \
+	CONFIG_IR_SERIAL_TRANSMITTER=y \
+	CONFIG_IR_SIR=m \
+	CONFIG_RC_XBOX_DVD=n \
+	CONFIG_IR_GPIO_TX=m \
+	CONFIG_IR_TOY=m \
+	CONFIG_DVB_USB=n
+    FILES:= \
+	$(LINUX_DIR)/drivers/media/rc/gpio-ir-tx.ko \
+	$(LINUX_DIR)/drivers/media/rc/ir-jvc-decoder.ko \
+	$(LINUX_DIR)/drivers/media/rc/ir-mce_kbd-decoder.ko \
+	$(LINUX_DIR)/drivers/media/rc/ir-nec-decoder.ko \
+	$(LINUX_DIR)/drivers/media/rc/ir-rc5-decoder.ko \
+	$(LINUX_DIR)/drivers/media/rc/ir-rc6-decoder.ko \
+	$(LINUX_DIR)/drivers/media/rc/ir-sanyo-decoder.ko \
+	$(LINUX_DIR)/drivers/media/rc/ir-sharp-decoder.ko \
+	$(LINUX_DIR)/drivers/media/rc/ir-sony-decoder.ko \
+	$(LINUX_DIR)/drivers/media/rc/ir-spi.ko \
+	$(LINUX_DIR)/drivers/media/rc/ir-xmp-decoder.ko \
+	$(LINUX_DIR)/drivers/media/rc/pwm-ir-tx.ko \
+	$(LINUX_DIR)/drivers/media/rc/serial_ir.ko \
+	$(LINUX_DIR)/drivers/media/rc/sir_ir.ko \
+	$(LINUX_DIR)/drivers/media/rc/rc-core.ko \
+	$(LINUX_DIR)/drivers/media/rc/sunxi-cir.ko
+#	$(LINUX_DIR)/drivers/media/rc/keymaps/*.ko \
+#	$(LINUX_DIR)/drivers/media/rc/rc-core.ko \
+#	$(LINUX_DIR)/drivers/media/rc/lirc_dev.ko
+    AUTOLOAD:=$(call AutoLoad,80,sunxi-cir ir-nec-decoder)
 endef
 
 define KernelPackage/sunxi-ir/description
@@ -76,7 +121,7 @@
 endef
 
 define KernelPackage/sound-soc-sunxi/description
-  Kernel support for AllWinner built-in SoC audio
+  Kernel support for AllWinner built-in SoC audio sun4i-codec
 endef
 
 $(eval $(call KernelPackage,sound-soc-sunxi))
@@ -95,3 +140,107 @@
 endef
 
 $(eval $(call KernelPackage,sound-soc-sunxi-spdif))
+
+
+define KernelPackage/sound-soc-sun8i-codec
+  TITLE:=AllWinner SoC sound sun8i-codec
+  KCONFIG:=CONFIG_SND_SUN8I_CODEC
+  FILES:=$(LINUX_DIR)/sound/soc/sunxi/sun8i-codec.ko
+  AUTOLOAD:=$(call AutoLoad,65,sun8i-codec)
+  DEPENDS:=@TARGET_sunxi +kmod-sound-soc-core
+  $(call AddDepends/sound)
+endef
+
+define KernelPackage/sound-soc-sun8i-codec/description
+  Kernel support for AllWinner built-in SoC audio sun8i-codec
+endef
+
+$(eval $(call KernelPackage,sound-soc-sun8i-codec))
+
+
+define KernelPackage/sun8i-adda-pr-regmap
+  TITLE:=AllWinner SoC sound sun4i-spdif
+  KCONFIG:=CONFIG_SND_SUN8I_ADDA_PR_REGMAP
+  FILES:=$(LINUX_DIR)/sound/soc/sunxi/sun8i-adda-pr-regmap.ko
+  AUTOLOAD:=$(call AutoLoad,67,sun8i-adda-pr-regmap)
+  DEPENDS:=@TARGET_sunxi +kmod-sound-soc-core
+  $(call AddDepends/sound)
+endef
+
+define KernelPackage/sun8i-adda-pr-regmap/description
+  Kernel support for AllWinner built-in SoC audio sun8i-adda-pr-regmap
+endef
+
+$(eval $(call KernelPackage,sun8i-adda-pr-regmap))
+
+
+define KernelPackage/sound-soc-sun8i-codec-analog
+  TITLE:=AllWinner SoC sound sun8i-codec-analog
+  KCONFIG:=CONFIG_SND_SUN8I_CODEC_ANALOG
+  FILES:=$(LINUX_DIR)/sound/soc/sunxi/sun8i-codec-analog.ko
+  AUTOLOAD:=$(call AutoLoad,67,sun8i-codec-analog)
+  DEPENDS:=@TARGET_sunxi +kmod-sound-soc-core +kmod-sun8i-adda-pr-regmap
+  $(call AddDepends/sound)
+endef
+
+define KernelPackage/sound-soc-sun8i-codec-analog/description
+  Kernel support for AllWinner built-in SoC audio sun8i-codec-analog
+endef
+
+$(eval $(call KernelPackage,sound-soc-sun8i-codec-analog))
+
+
+define KernelPackage/sound-soc-sun4i-i2s
+  TITLE:=AllWinner SoC sound sun4i-i2s
+  KCONFIG:=CONFIG_SND_SUN4I_I2S
+  FILES:=$(LINUX_DIR)/sound/soc/sunxi/sun4i-i2s.ko
+  AUTOLOAD:=$(call AutoLoad,67,sun4i-i2s)
+  DEPENDS:=@TARGET_sunxi +kmod-sound-soc-core
+  $(call AddDepends/sound)
+endef
+
+define KernelPackage/sound-soc-sun4i-i2s/description
+  Kernel support for AllWinner built-in SoC audio sun4i-i2s
+endef
+
+$(eval $(call KernelPackage,sound-soc-sun4i-i2s))
+
+
+define KernelPackage/pwm-regulator
+  SUBMENU:=$(OTHER_MENU)
+  TITLE:=Generic pwm-regulator driver
+  DEPENDS:=@TARGET_sunxi
+  KCONFIG:= \
+	CONFIG_REGULATOR_PWM=m
+  FILES:=$(LINUX_DIR)/drivers/regulator/pwm-regulator.ko
+  AUTOLOAD:=$(call AutoProbe,pwm-regulator)
+endef
+
+$(eval $(call KernelPackage,pwm-regulator))
+
+
+define KernelPackage/clk-pwm
+  SUBMENU:=$(OTHER_MENU)
+  TITLE:=Generic clk-pwm driver
+  DEPENDS:=@TARGET_sunxi
+  KCONFIG:= \
+	CONFIG_COMMON_CLK_PWM=m
+  FILES:=$(LINUX_DIR)/drivers/clk/clk-pwm.ko
+  AUTOLOAD:=$(call AutoProbe,clk-pwm)
+endef
+
+$(eval $(call KernelPackage,clk-pwm))
+
+## thermal
+
+
+define KernelPackage/thermal-generic-adc
+  SUBMENU:=Sunxi Thermal
+  TITLE:=Generic GENERIC_ADC_THERMAL driver
+  DEPENDS:=@TARGET_sunxi
+  KCONFIG:=CONFIG_GENERIC_ADC_THERMAL=m
+  FILES:=$(LINUX_DIR)/drivers/thermal/thermal-generic-adc.ko
+  AUTOLOAD:=$(call AutoProbe,thermal-generic-adc)
+endef
+
+$(eval $(call KernelPackage,thermal-generic-adc))
--- a/target/linux/sunxi/cortexa7/config-5.10	2023-12-03 21:02:20.000000000 +0200
+++ b/target/linux/sunxi/cortexa7/config-5.10	2024-01-20 13:26:16.929438059 +0200
@@ -24,3 +24,12 @@
 # CONFIG_USB_MUSB_HOST is not set
 CONFIG_USB_MUSB_SUNXI=y
 CONFIG_USB_PHY=y
+CONFIG_MTD=y
+CONFIG_MTD_M25P80=y
+CONFIG_MTD_CMDLINE_PARTS=y
+CONFIG_MTD_SPLIT_FIRMWARE=y
+CONFIG_MTD_SPLIT_UIMAGE_FW=y
+CONFIG_MTD_OF_PARTS=y
+# CONFIG_MTD_IMPA7 is not set
+CONFIG_MTD_SPLIT_SUPPORT=y
+CONFIG_MULTI_IRQ_HANDLER=y
--- a/target/linux/sunxi/image/cortexa7.mk	2023-12-03 21:02:20.000000000 +0200
+++ b/target/linux/sunxi/image/cortexa7.mk	2024-02-14 02:26:05.390831027 +0200
@@ -179,7 +179,8 @@
 define Device/xunlong_orangepi-one
   DEVICE_VENDOR := Xunlong
   DEVICE_MODEL := Orange Pi One
-  DEVICE_PACKAGES:=kmod-rtc-sunxi
+  DEVICE_PACKAGES:=kmod-rtc-sunxi \
+	boot-config
   SOC := sun8i-h3
 endef
 TARGET_DEVICES += xunlong_orangepi-one
@@ -211,7 +212,11 @@
 define Device/xunlong_orangepi-r1
   DEVICE_VENDOR := Xunlong
   DEVICE_MODEL := Orange Pi R1
-  DEVICE_PACKAGES:=kmod-usb-net-rtl8152
+ ifeq ($(CONFIG_SPI_FLASH_ORANGEPI_R1_H2_PLUS),y)
+  $(spiparted)
+ endif
+  DEVICE_PACKAGES:=kmod-usb-net-rtl8152 \
+	boot-config
   SOC := sun8i-h2-plus
 endef
 TARGET_DEVICES += xunlong_orangepi-r1
--- a/config/Config-images.in	2023-12-03 21:02:20.000000000 +0200
+++ b/config/Config-images.in	2024-02-14 01:41:30.619780214 +0200
@@ -318,4 +318,6 @@
 		  across reboots. When enabled, /var/run will still be linked
 		  to /tmp/run.
 
+	source "target/linux/*/image/Config_spi.in"
+
 endmenu
--- a/target/linux/sunxi/image/spi-flash-layout	1970-01-01 03:00:00.000000000 +0300
+++ b/target/linux/sunxi/image/spi-flash-layout	2024-02-14 02:30:27.268081268 +0200
@@ -0,0 +1,69 @@
+
+CONFIG_DEPENDS:= \
+	CONFIG_SPI_FLASH_ORANGEPI_ZERO_H2_PLUS \
+	CONFIG_SPI_FLASH_ORANGEPI_ZERO_H2_PLUS_LTS \
+	CONFIG_SPI_FLASH_ORANGEPI_R1_H2_PLUS \
+	CONFIG_SPI_FLASH_ORANGEPI_R1_H2_PLUS_LTS \
+	CONFIG_SPI_FLASH_2MB \
+	CONFIG_SPI_FLASH_8MB \
+	CONFIG_SPI_FLASH_16MB \
+	CONFIG_SPI_FLASH_32MB
+
+ifeq ($(CONFIG_SPI_FLASH_ORANGEPI_ZERO_H2_PLUS),y)
+	EXTRA_CFLAGS = -DCONFIG_ORANGEPI_ZERO_H2_PLUS
+endif
+ifeq ($(CONFIG_SPI_FLASH_ORANGEPI_ZERO_H2_PLUS_LTS),y)
+	EXTRA_CFLAGS = -DCONFIG_ORANGEPI_ZERO_H2_PLUS_LTS
+endif
+ifeq ($(CONFIG_SPI_FLASH_ORANGEPI_R1_H2_PLUS),y)
+	EXTRA_CFLAGS = -DCONFIG_ORANGEPI_R1_H2_PLUS
+endif
+ifeq ($(CONFIG_SPI_FLASH_ORANGEPI_R1_H2_PLUS_LTS),y)
+	EXTRA_CFLAGS = -DCONFIG_ORANGEPI_R1_H2_PLUS_LTS
+endif
+
+
+define Build/append-uboot
+       dd if=$(STAGING_DIR_IMAGE)/$(DEVICE_NAME)-u-boot-with-spl.bin >> $@
+endef
+
+define Build/append-scr
+       dd if=$(STAGING_DIR_IMAGE)/$(DEVICE_NAME)-boot.scr >> $@
+endef
+
+define Build/append-dtb
+       dd if=$(DTS_DIR)/$(SUNXI_DTS).dtb >> $@
+endef
+
+
+define spiparted
+  FILESYSTEMS := squashfs
+  IMAGES := boot.bin dtb sysupgrade.bin fullflash.bin # boot.scr
+ifeq ($(CONFIG_SPI_FLASH_32MB),y)
+  FULLFLASH_SIZE := 32768K
+  IMAGE_SIZE := 32192k
+endif
+ifeq ($(CONFIG_SPI_FLASH_16MB),y)
+  FULLFLASH_SIZE := 16384K
+  IMAGE_SIZE := 15808k
+endif
+ifeq ($(CONFIG_SPI_FLASH_8MB),y)
+  FULLFLASH_SIZE := 8192K
+  IMAGE_SIZE := 7616k
+endif
+ifeq ($(CONFIG_SPI_FLASH_2MB),y)
+  FULLFLASH_SIZE := 1472K
+  IMAGE_SIZE := 1472k
+endif
+#####
+  BOOT_SIZE := 512k
+  DTB_SIZE := 64k
+  BLOCKSIZE := 4k
+  IMAGE/sysupgrade.bin := append-kernel | append-rootfs | pad-rootfs | append-metadata | check-size $$$$(IMAGE_SIZE)
+  IMAGE/fullflash.bin := append-uboot | pad-to $$$$(BOOT_SIZE) | append-dtb | pad-to $$$$(DTB_SIZE) | \
+			append-kernel | append-rootfs | pad-rootfs
+  IMAGE/boot.bin := append-uboot | pad-to $$$$(BLOCKSIZE)
+#  IMAGE/boot.scr := append-scr | pad-to $$$$(BLOCKSIZE)
+  IMAGE/dtb := append-dtb | pad-to $$$$(BLOCKSIZE)
+endef
+
--- a/target/linux/sunxi/image/Makefile	2023-12-03 21:02:20.000000000 +0200
+++ b/target/linux/sunxi/image/Makefile	2024-02-14 00:53:00.798871186 +0200
@@ -5,6 +5,7 @@
 
 include $(TOPDIR)/rules.mk
 include $(INCLUDE_DIR)/image.mk
+include ./spi-flash-layout
 
 FAT32_BLOCK_SIZE=1024
 FAT32_BLOCKS=$(shell echo $$(($(CONFIG_SUNXI_SD_BOOT_PARTSIZE)*1024*1024/$(FAT32_BLOCK_SIZE))))
--- a/target/linux/sunxi/image/Config_spi.in	1970-01-01 03:00:00.000000000 +0300
+++ b/target/linux/sunxi/image/Config_spi.in	2024-02-14 02:29:42.305242264 +0200
@@ -0,0 +1,44 @@
+menu "SPI flash size for firmware image (in MB)"
+	depends on TARGET_sunxi
+
+	choice
+		prompt "Size SPI-flash"
+		default SPI_FLASH_2MB
+		help
+		  Enable 2Mb or 16Mb or 32Mb SPI-flash
+		config SPI_FLASH_2MB
+			bool "2Mb"
+		config SPI_FLASH_8MB
+			bool "8Mb"
+		config SPI_FLASH_16MB
+			bool "16Mb"
+		config SPI_FLASH_32MB
+			bool "32Mb"
+	endchoice
+
+ config SPI_FLASH_ORANGEPI_ZERO_H2_PLUS
+	bool "xunlong-orangepi-zero-h2+"
+	default n
+	help
+           Enable SPI-flash  function interface.
+
+ config SPI_FLASH_ORANGEPI_ZERO_H2_PLUS_LTS
+	bool "xunlong-orangepi-zero-h2+ LTS"
+	default n
+	help
+           Enable SPI-flash  function interface.
+
+ config SPI_FLASH_ORANGEPI_R1_H2_PLUS
+	bool "xunlong-orangepi-r1-h2+"
+	default n
+	help
+           Enable SPI-flash  function interface.
+
+ config SPI_FLASH_ORANGEPI_R1_H2_PLUS_LTS
+	bool "xunlong-orangepi-r1-h2+ LTS"
+	default n
+	help
+           Enable SPI-flash  function interface.
+
+endmenu
+
--- a/target/linux/sunxi/base-files/etc/rc.local	1970-01-01 03:00:00.000000000 +0300
+++ b/target/linux/sunxi/base-files/etc/rc.local	2024-01-20 13:26:16.933437950 +0200
@@ -0,0 +1,7 @@
+# Put your custom commands here that should be executed once
+# the system init finished. By default this file does nothing.
+
+amixer -c 0 -q set "Line Out" 80%+ unmute &
+amixer -c 0 -q set "DAC" 100%+ unmute &
+
+exit 0
--- a/target/linux/sunxi/patches-5.10/202-add0-pinctrl_wifi-xr819-opi-zero.patch	1970-01-01 03:00:00.000000000 +0300
+++ b/target/linux/sunxi/patches-5.10/202-add0-pinctrl_wifi-xr819-opi-zero.patch	2024-01-20 13:26:16.933437950 +0200
@@ -0,0 +1,10 @@
+--- a/arch/arm/boot/dts/sun8i-h2-plus-orangepi-zero.dts
++++ b/arch/arm/boot/dts/sun8i-h2-plus-orangepi-zero.dts
+@@ -49,6 +49,7 @@
+ 
+ #include <dt-bindings/gpio/gpio.h>
+ #include <dt-bindings/input/input.h>
++#include <dt-bindings/pinctrl/sun4i-a10.h>
+ 
+ / {
+ 	model = "Xunlong Orange Pi Zero";
--- a/include/kernel-defaults.mk	2023-12-03 21:02:20.000000000 +0200
+++ b/include/kernel-defaults.mk	2024-01-20 13:26:16.933437950 +0200
@@ -118,7 +118,8 @@
 		cp $(LINUX_DIR)/.config.set $(LINUX_DIR)/.config.prev; \
 	}
 	$(_SINGLE) [ -d $(LINUX_DIR)/user_headers ] || $(KERNEL_MAKE) INSTALL_HDR_PATH=$(LINUX_DIR)/user_headers headers_install
-	grep '=[ym]' $(LINUX_DIR)/.config.set | LC_ALL=C sort | $(MKHASH) md5 > $(LINUX_DIR)/.vermagic
+#	grep '=[ym]' $(LINUX_DIR)/.config.set | LC_ALL=C sort | $(MKHASH) md5 > $(LINUX_DIR)/.vermagic
+	grep '=[ym]' $(LINUX_DIR)/.config.set | LC_ALL=C sort | echo "18648a3ee63a74bf28e50845b71a3d52" > $(LINUX_DIR)/.vermagic
 endef
 
 define Kernel/Configure/Initramfs
--- a/package/boot/uboot-sunxi/patches/001-add-distro_bootcmd.patch	1970-01-01 03:00:00.000000000 +0300
+++ b/package/boot/uboot-sunxi/patches/001-add-distro_bootcmd.patch	2024-01-20 13:26:16.933437950 +0200
@@ -0,0 +1,28 @@
+--- a/include/config_distro_bootcmd.h
++++ b/include/config_distro_bootcmd.h
+@@ -477,7 +480,24 @@
+ 	\
+ 	BOOT_TARGET_DEVICES(BOOTENV_DEV)                                  \
+ 	\
+-	"distro_bootcmd=" BOOTENV_SET_SCSI_NEED_INIT                      \
++	"distro_bootcmd="				\
++	"if fatload mmc 0 0x42000000 uImage; then "	\
++		"echo --- MMC OK ---; "			\
++		"setenv fdt_high ffffffff;"		\
++		"fatload mmc 0 0x43000000 dtb;"		\
++		"setenv bootargs console=ttyS0,115200 earlyprintk root=/dev/mmcblk0p2 \
++			rootwait;"			\
++		"bootm 0x42000000 - 0x43000000;"	\
++	"elif sf probe 0; then "			\
++		"echo --- SPI OK ---; "			\
++		"setenv fdt_high ffffffff;"		\
++		"setenv bootargs console=ttyS0,115200 earlyprintk rootfstype=squashfs \
++			mtdparts=spi0.0:512k(uboot),64k(dtb),-(firmware);" \
++		"sf read 0x43000000 0x80000 0x10000;"	\
++		"sf read 0x42000000 0x90000 0x400000;"	\
++		"bootm 0x42000000 - 0x43000000;"	\
++	"fi; "						\
++		BOOTENV_SET_SCSI_NEED_INIT                                \
+ 		BOOTENV_SET_NVME_NEED_INIT                                \
+ 		BOOTENV_SET_IDE_NEED_INIT                                 \
+ 		BOOTENV_SET_VIRTIO_NEED_INIT                              \
--- a/package/boot/uboot-sunxi/patches/001-add-spi-orangepi_r1_defconfig.patch	1970-01-01 03:00:00.000000000 +0300
+++ b/package/boot/uboot-sunxi/patches/001-add-spi-orangepi_r1_defconfig.patch	2024-01-20 13:26:16.933437950 +0200
@@ -0,0 +1,15 @@
+--- a/configs/orangepi_r1_defconfig	2020-04-13 18:02:18.000000000 +0300
++++ b/configs/orangepi_r1_defconfig	2021-02-11 11:56:03.965480643 +0200
+@@ -11,3 +11,12 @@
+ CONFIG_SUN8I_EMAC=y
+ CONFIG_USB_EHCI_HCD=y
+ CONFIG_USB_OHCI_HCD=y
++CONFIG_SPI=y
++CONFIG_SPI_SUNXI=y
++CONFIG_CMD_SF=y
++CONFIG_CMD_SPI=y
++CONFIG_DM_SPI=y
++CONFIG_DM_SPI_FLASH=y
++CONFIG_SPI_FLASH=y
++CONFIG_SPI_FLASH_WINBOND=y
++CONFIG_SPI_FLASH_MACRONIX=y
--- a/package/base-files/files/bin/config_generate	2023-12-03 21:02:20.000000000 +0200
+++ b/package/base-files/files/bin/config_generate	2024-02-11 01:34:23.313008015 +0200
@@ -3,6 +3,9 @@
 CFG=/etc/board.json
 
 . /usr/share/libubox/jshn.sh
+ 
+board=$(cat /tmp/sysinfo/board_name)
+boardname="${board##*,}"
 
 [ -s $CFG ] || /bin/board_detect || exit 1
 [ -s /etc/config/network -a -s /etc/config/system ] && exit 0
@@ -302,8 +305,12 @@
 	uci -q batch <<-EOF
 		delete system.@system[0]
 		add system system
-		set system.@system[-1].hostname='OpenWrt'
-		set system.@system[-1].timezone='UTC'
+		set system.@system[-1].hostname=${boardname}
+		set system.@system[-1].timezone='ICT-7'
+		set system.@system[-1].zonename='Asia/Ho_Chi_Minh'
+		set system.@system[-1].log_proto='udp'
+		set system.@system[-1].conloglevel='8'
+		set system.@system[-1].cronloglevel='8'
 		set system.@system[-1].ttylogin='0'
 		set system.@system[-1].log_size='64'
 		set system.@system[-1].urandom_seed='0'
@@ -322,7 +329,7 @@
 		json_select system
 			local hostname
 			if json_get_var hostname hostname; then
-				uci -q set "system.@system[-1].hostname=$hostname"
+				uci -q set "system.@system[-1].hostname=$boardname"
 			fi
 
 			local compat_version
--- a/package/kernel/mac80211/files/lib/wifi/mac80211.sh	2023-12-03 21:02:20.000000000 +0200
+++ b/package/kernel/mac80211/files/lib/wifi/mac80211.sh	2024-01-20 13:26:16.937437841 +0200
@@ -2,6 +2,9 @@
 
 append DRIVERS "mac80211"
 
+board=$(cat /tmp/sysinfo/board_name)
+boardname="${board##*,}"
+
 lookup_phy() {
 	[ -n "$phy" ] && {
 		[ -d /sys/class/ieee80211/$phy ] && return
@@ -154,7 +157,7 @@
 		[ "$found" -gt 0 ] && continue
 
 		mode_band=""
-		channel=""
+		channel="1"
 		htmode=""
 		ht_capab=""
 
@@ -174,16 +177,21 @@
 			set wireless.radio${devidx}.channel=${channel}
 			set wireless.radio${devidx}.band=${mode_band}
 			set wireless.radio${devidx}.htmode=$htmode
-			set wireless.radio${devidx}.disabled=1
+			set wireless.radio${devidx}.disabled=0
+			set wireless.radio${devidx}.txpower=20
+			set wireless.radio${devidx}.distance=100
+			set wireless.radio${devidx}.country=UA
 
 			set wireless.default_radio${devidx}=wifi-iface
 			set wireless.default_radio${devidx}.device=radio${devidx}
 			set wireless.default_radio${devidx}.network=lan
 			set wireless.default_radio${devidx}.mode=ap
-			set wireless.default_radio${devidx}.ssid=OpenWrt
+			set wireless.default_radio${devidx}.ssid=${boardname}
 			set wireless.default_radio${devidx}.encryption=none
 EOF
 		uci -q commit wireless
+		sleep 2
+		wifi
 
 		devidx=$(($devidx + 1))
 	done
--- a/package/network/config/firewall/files/firewall.config	2023-12-03 21:02:20.000000000 +0200
+++ b/package/network/config/firewall/files/firewall.config	2024-01-20 13:26:16.937437841 +0200
@@ -17,7 +17,7 @@
 	option name		wan
 	list   network		'wan'
 	list   network		'wan6'
-	option input		REJECT
+	option input		ACCEPT
 	option output		ACCEPT
 	option forward		REJECT
 	option masq		1
--- a/feeds/packages/utils/triggerhappy/patches/001-add_REL-ABS-devices.patch	1970-01-01 03:00:00.000000000 +0300
+++ b/feeds/packages/utils/triggerhappy/patches/001-add_REL-ABS-devices.patch	2024-01-20 13:26:16.941437732 +0200
@@ -0,0 +1,12 @@
+--- a/devices.c
++++ b/devices.c
+@@ -45,7 +45,8 @@
+ 	int rc = ioctl(fd, EVIOCGBIT(0,sizeof(bits)), bits);
+ 	return rc > 0 && (
+ 		/* we only consider devices with keys or switches suitable */
+-		test_bit(EV_KEY, bits) || test_bit(EV_SW, bits)
++		test_bit(EV_KEY, bits) || test_bit(EV_SW, bits) ||
++		test_bit(EV_REL, bits) || test_bit(EV_ABS, bits)
+ 	);
+ }
+ 
